java/src/main/native/src/RmmJni.cpp

```diff
@@ -14,14 +14,118 @@
  * limitations under the License.
  */
 
+#include <rmm/mr/device/default_memory_resource.hpp>
+
 #include "jni_utils.hpp"
 
+namespace {
+
+class java_event_handler_memory_resource : public rmm::mr::device_memory_resource {
+public:
+  java_event_handler_memory_resource(JNIEnv* env, jobject jhandler) {
+    if (env->GetJavaVM(&jvm) < 0) {
+      throw cudf::jni::jni_exception("GetJavaVM failed");
+    }
+    jclass cls = env->GetObjectClass(jhandler);
+    if (cls == nullptr) {
+      throw cudf::jni::jni_exception("class not found");
+    }
+    on_alloc_fail_method = env->GetMethodID(cls, "onAllocFailure", "(J)Z");
+    if (on_alloc_fail_method == nullptr) {
+      throw cudf::jni::jni_exception("onAllocFailure method");
+    }
+    handler_obj = env->NewGlobalRef(jhandler);
+    if (handler_obj == nullptr) {
+      throw cudf::jni::jni_exception("global ref");
+    }
+
+    // The previous resource could be null, but getting the current resource never is.
+    resource = rmm::mr::get_default_resource();
+    rmm::mr::device_memory_resource* old = rmm::mr::set_default_resource(this);
+    if (old != nullptr) {
+      resource = old;
+    }
+  }
+
+  virtual ~java_event_handler_memory_resource() {
+    // TODO: This is problematic if the memory manager was updated in the interim.
+    rmm::mr::set_default_resource(resource);
```

Comment:
See above comments about making upstream a parameter and lifting the responsibility for setting the default resource outside of this class. I likewise think resetting the default resource should be lifted outside of this class, because I agree with your comment that this is problematic.

Reply:
I don't have a great answer for how to deal with concurrent modification of the memory resource.  I tried to add some code to at least detect it and throw an exception if it happens, but it probably isn't doing the right thing in all situations to recover from it.